### PR TITLE
feature: Serialization + deserialization of viewer state

### PR DESCRIPTION
**Pull request by ShrimpCryptid**
Second part of progress towards #107. Adds serialization and deserialization of the new ViewerState introduced by @frasercl and associated unit tests.

*Estimated review size: medium, ~30 minutes. (~300 lines are unit tests.)*

## Changes
- Adds methods for serializing and deserializing the ViewerState type.
- Adds a method for serializing the ChannelState.
  - Currently missing handling for control points-- pending a PR @frasercl is working on
- Adds unit tests for serializing/deserializing viewer state types.

### CODE REVIEW

File: src/aics-image-viewer/shared/enums.ts
Comment (ShrimpCryptid) [inline src/aics-image-viewer/shared/enums.ts:1]: Allows the `parseStringEnum` function to work. `const` enums are removed at runtime, while non-`const` enums persist as objects.

File: src/aics-image-viewer/shared/enums.ts [hunk 1]
```diff
@@ -1,17 +1,17 @@
-export const enum ViewMode {
+export enum ViewMode {
```
Comment (ShrimpCryptid) [inline src/aics-image-viewer/shared/enums.ts:1]: Allows the `parseStringEnum` function to work. `const` enums are removed at runtime, while non-`const` enums persist as objects.

File: website/utils/url_utils.ts [hunk 2]
```diff
@@ -26,15 +30,21 @@ const SLICE_REGEX = /^[0-9.]*,[0-9.]*,[0-9.]*$/;
...
+// Classes have default `undefined` values for all properties
+// to ensure the keys exist when the object is created with no arguments.
+// This allows us to iterate over the keys and turn them into a string list of allowed keys.
+/**
+ * The serialized form of a ViewerChannelSetting, as a dictionary object.
+ */
+export class ViewerChannelSettingParams {
   /** Color, as a 6-digit hex color.  */
-  col?: string;
+  col?: string = undefined;
```
Comment (meganrm) [inline website/utils/url_utils.ts:41]: suggestion: make these all enum values so you can do `[urlParam.Color]: string = undefined` to make it more readable

File: website/utils/url_utils.ts [hunk 3]
```diff
@@ -320,66 +386,92 @@ export function deserializeViewerChannelSetting(
...
-  args.imageDownloadHref = datasetData.downloadRoot + "/" + id;
-  // args.fovDownloadHref = datasetData.downloadRoot + "/" + id;
-
-  const fileInfo = await db.getFileInfoByCellId(id);
-  args.imageUrl = baseUrl + fileInfo!.volumeviewerPath;
-  args.parentImageUrl = baseUrl + fileInfo!.fovVolumeviewerPath;
-
-  return args;
+export function serializeViewerChannelSetting(channelSetting: ChannelState): ViewerChannelSettingParams {
+  return removeUndefinedProperties({
```
Comment (ShrimpCryptid) [inline website/utils/url_utils.ts:390]: @ShrimpCryptid do you need this here? Is `colorizeAlpha` ever undefined?

File: website/utils/url_utils.ts [hunk 4]
```diff
@@ -53,59 +63,114 @@ export type ViewerChannelSettingJson = {
...
+const isParamKey = (key: string): key is keyof Params => allowedParamKeys.indexOf(key as keyof Params) !== -1;
+const isChannelKey = (key: string): key is keyof ChannelParams => CHANNEL_STATE_KEY_REGEX.test(key);
 
+/**
+ * Filters a set of URLSearchParams for only the keys that are valid parameters for the viewer.
+ * Non-matching keys are discarded.
+ * @param searchParams Input URL search parameters.
+ * @returns a dictionary object matching the type of `Params`.
+ */
 export function urlSearchParamsToParams(searchParams: URLSearchParams): Params {
```
Comment (meganrm) [inline website/utils/url_utils.ts:170]: nit but this name confuses me. maybe `urlSearchParamsToAllowableParams` or `getAllowedParams` or `cleanUrlSearchParams`. I think `Params` as a type could also be changes to `AppParams` or `AllowableParams`

File: website/utils/url_utils.ts (hunk 3 above)
Comment (meganrm) [inline website/utils/url_utils.ts:390]: asked in person: is this necessary?

File: website/utils/url_utils.ts [hunk 5]
```diff
@@ -320,66 +386,92 @@ export function deserializeViewerChannelSetting(
...
-export async function getArgsFromParams(urlSearchParams: URLSearchParams): Promise<{
-  args: Partial<AppProps>;
-  viewerSettings: Partial<ViewerState>;
-}> {
-  const params = urlSearchParamsToParams(urlSearchParams);
-  let args: Partial<AppProps> = {};
-  const viewerSettings: Partial<ViewerState> = {};
+export function deserializeViewerState(params: ViewerStateParams): Partial<ViewerState> {
+  const result: Partial<ViewerState> = {
+    maskAlpha: parseStringInt(params.mask, 0, 100),
```
Comment (meganrm) [inline website/utils/url_utils.ts:406]: suggestion for readability: create a general `parse` function that takes a parameter, a type and options, that then calls each of these functions. 
`maskAlpha: parse(params.mask, paramType.int, [0, 100])`

or a lookup table from the urlParam enum to the state field with options

**Review by meganrm — APPROVED** [review]
I left some comments for improving clarity. If you make significant changes based on those comments, you can re-request a review so I can see what you ended up doing

**Review by interim17 — APPROVED** [review]

**Review by rugeli — APPROVED** [review]
sorry for taking so long, look good to me!

**Comment by ShrimpCryptid** [comment]
Thanks, y'all! I'm going to work on making some of the changes Megan suggested (probably should have marked it as draft while that was happening). Appreciate the eyes on this PR though :')

File: website/utils/url_utils.ts (hunk 5 above)
Comment (ShrimpCryptid) [reply website/utils/url_utils.ts:406]: Tried noodling on this with Cameron... The main blocker was trying to get the same functionality we currently have when parsing enums.  Currently, the `parseStringEnum()` function really nicely guarantees that the returned type is of the enum that was passed in. 

There isn't an easy way with TypeScript to make `parse` generic *and* still support the correct return type for enums passed in as variables.

**Review by meganrm — APPROVED** [review]

**Pull request merged**